elgato.dart

```diff
@@ -1,81 +1,87 @@
 import 'dart:async';
 import 'dart:convert' as convert;
 import 'dart:io';
+
 import 'package:dcli/dcli.dart';
 import 'package:http/http.dart' as http;
 import 'package:multicast_dns/multicast_dns.dart';
 
-// We store the results of the IP lookup in a file for future speedups
+// Store the results of the IP lookup locally to speed things up
 final cacheFile = join(HOME, ".elgato.dart.cache");
 
-String lightIpPort = "";
-Uri url = Uri.parse("http://$lightIpPort/elgato/lights");
+var lightIpPorts = [];
+var urls = lightIpPorts.map((x) => Uri.parse('http://$x/elgato/lights'));
 
 /// Discover Elgato lights on the network
 findLights() async {
   final String svcName = "_elg";
   final MDnsClient client = MDnsClient();
   await client.start();
 
-  await for (PtrResourceRecord ptr in client
-      .lookup<PtrResourceRecord>(ResourceRecordQuery.serverPointer(svcName))) {
-    await for (SrvResourceRecord srv in client.lookup<SrvResourceRecord>(
-        ResourceRecordQuery.service(ptr.domainName))) {
-      await for (IPAddressResourceRecord ip
-          in client.lookup<IPAddressResourceRecord>(
-              ResourceRecordQuery.addressIPv4(srv.target))) {
-        lightIpPort = '${ip.address.host}:${srv.port}';
+  var svcPointer = ResourceRecordQuery.serverPointer(svcName);
+  await for (var ptr in client.lookup<PtrResourceRecord>(svcPointer)) {
+    var svcQuery = ResourceRecordQuery.service(ptr.domainName);
+    await for (var srv in client.lookup<SrvResourceRecord>(svcQuery)) {
+      var svcName = ResourceRecordQuery.addressIPv4(srv.target);
+      await for (var ip in client.lookup<IPAddressResourceRecord>(svcName)) {
+        lightIpPorts.add('${ip.address.host}:${srv.port}');
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏it appears that the material change here is that we've gone from `lightIpPort = ...` to. `lightIpPorts.add(...` - is that correct? Is there a specific reason for the other changes, or is it cosmetic?

Reply:
Yep, that's the key change. ``lightIpPort`` was previously a string, only storing one light's IP and port. So I replaced that to a list. This also meant I needed to rewrite a bit around the cache file logic further down.

Anyway, the rest of the changes in this particular segment are indeed entirely cosmetic. As per Dart's style guide it's recommended to not write out the full class names where Dart can discover them automatically (and there's no ambiguity). I find it helps a lot with readability, and since I initially thought there was an error in the network lookup I rewrote it like this to more easily understand what was going on.

If you want to rewrote those lines I don't have any super strong objections (but not sure how to actually do that in practice? do I need to redo the pull request?)

Reply:
Thanks for the explanation! That makes a lot of sense

> As per Dart's style guide it's recommended to not write out the full class names where Dart can discover them automatically...

I went looking for this in the [dart style guide](https://dart.dev/guides/language/effective-dart/style) - do you have a link/reference I can read? I'm definitely in support for using automatic type inference, but for some reason in my brain I was supposed to write it this way. If that's changed in the Dart ecosystem, I'm happy to adapt!

No need to rewrite yet - if it came to that, there's always the option to add a fresh commit restoring the changes, and then I can squash it all to a single commit on merge, if that's how we want to go.

Reply:
I stand corrected! Could swear this was in the style guide, I've been doing it for a while myself.

The only official reference I can find is this section https://dart.dev/guides/language/effective-dart/design#type-inference and ```Do annotate when inference doesn’t have enough context, even when dynamic is the type you want.```.

I'd argue that for this particular segment, we do have enough context, so the types aren't needed.

Reply:
I agree that this segment has enough context to not need further annotations - specifically since they are all declared and used within the same nested for loops. Will test on my computer later on tonight.

This does bring up a new challenge, if you're interested - now if there's a wireless network with more than one office (like a home office with two people) and both have Elgato lights, this tool will likely flip all of them. Not something I'm currently worried about, but something to consider!